### PR TITLE
feat: add mutations abstraction to deal with mnesia updates

### LINKED ISSUE #331: volatile sync error on get_events
```
mainnet_1           | 21:37:15.070 [error] Tx {aetx,contract_call_tx,aect_call_tx,104,{contract_call_tx,{id,account,<<127,19,38,206,133,114,28,107,114,242,140,75,21,37,155,45,79,211,114,88,64,95,96,195,107,184,57,112,251,74,176,59>>},569,{id,contract,<<130,39,65,134,89,4,253,113,11,203,0,143,186,176,170,86,114,216,110,120,189,213,139,155,63,155,244,104,230,72,51,165>>},3,300000000000000,0,0,5810000,100000000000,<<43,17,214,154,197,85,63>>,[],<<127,19,38,206,133,114,28,107,114,242,140,75,21,37,155,45,79,211,114,88,64,95,96,195,107,184,57,112,251,74,176,59>>}} cannot be applied due to an error {exit,{aborted,{cyclic,aeternity@localhost,{aec_chain_state,top_block_node},read,read,{tid,148502924,<0.922.0>}}}}
mainnet_1           | 21:37:15.071 [error] Supervisor 'Elixir.AeMdw.Db.Sync.Supervisor' had child 'Elixir.AeMdw.Db.Sync' started with 'Elixir.AeMdw.Db.Sync':start_link([]) at <0.1533.0> exit with reason no match of right hand value {aborted,{{badmatch,{error,{exit,{aborted,{cyclic,aeternity@localhost,{aec_chain_state,top_block_node},read,read,{tid,148502924,<0.922.0>}}}}}},[{'Elixir.AeMdw.Contract',get_events,1,[{file,"lib/ae_mdw/contract.ex"},{line,347}]},{'Elixir.AeMdw.Contract',get_grouped_events,1,[{file,"lib/ae_mdw/contract.ex"},{line,328}]},{'Elixir.AeMdw.Db.Sync.Transaction',sync_micro_block,2,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,109}]},{'Elixir.Enum','-reduce/3-lists^foldl/2-0-',3,[{file,"lib/enum..."},...]},...]}} in 'Elixir.AeMdw.Db.Sync.Transaction':sync_generation/2 line 60 in context child_terminated
mainnet_1           | 21:37:15.072 [error] Error in process <0.7590.461> on node aeternity@localhost with exit value:
mainnet_1           | {{badmatch,{aborted,{{badmatch,{error,{exit,{aborted,{cyclic,aeternity@localhost,{aec_chain_state,top_block_node},read,read,{tid,148502924,<0.922.0>}}}}}},[{'Elixir.AeMdw.Contract',get_events,1,[{file,"lib/ae_mdw/contract.ex"},{line,347}]},{'Elixir.AeMdw.Contract',get_grouped_events,1,[{file,"lib/ae_mdw/contract.ex"},{line,328}]},{'Elixir.AeMdw.Db.Sync.Transaction',sync_micro_block,2,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,109}]},{'Elixir.Enum','-reduce/3-lists^foldl/2-0-',3,[{file,"lib/enum.ex"},{line,2111}]},{'Elixir.AeMdw.Db.Sync.Transaction','-sync_generation/2-fun-1-',4,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,78}]},{mnesia_tm,apply_fun,3,[{file,"mnesia_tm.erl"},{line,842}]},{mnesia_tm,execute_transaction,5,[{file,"mnesia_tm.erl"},{line,818}]},{'Elixir.AeMdw.Db.Sync.Transaction',sync_generation,2,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,61}]},{'Elixir.AeMdw.Db.Sync.Transaction',sync,3,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,42}]},{'Elixir.AeMdw.Db.Sync',run_action,1,[{file,"lib/ae_mdw/db/sync/sync.ex"},{line,99}]}]}}},[{'Elixir.AeMdw.Db.Sync.Transaction',sync_generation,2,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,60}]},{'Elixir.AeMdw.Db.Sync.Transaction',sync,3,[{file,"lib/ae_mdw/db/sync/transaction.ex"},{line,42}]},{'Elixir.AeMdw.Db.Sync',run_action,1,[{file,"lib/ae_mdw/db/sync/sync.ex"},{line,99}]}]}
mainnet_1           | 21:37:15.072 [error] Supervisor 'Elixir.AeMdw.Db.Sync.Supervisor' had child 'Elixir.AeMdw.Db.Sync' started with 'Elixir.AeMdw.Db.Sync':start_link([]) at <0.1533.0> exit with reason reached_max_restart_intensity in context shutdown
```

By the log, the point it aborts is in line:
`{:ok, _, _, events} = :aec_block_micro_candidate.apply_block_txs_strict(txs, trees_in, env)`

at `Elixir.AeMdw.Contract',get_events,1,[{file,"lib/ae_mdw/contract.ex"},{line,347}]}`

### PR DESCRIPTION
**Pull request by sborrazas**
The goal of this change is to be able to isolate mnesia updates into
mutations which will all be ran in a single transaction and in the
least amount of time possible.

By making first-class mutations we can also debug easily what changes
any given block or transaction makes.

This also fixes random errors that come up inside mnesia transactions:
mnesia transactions do not work properly when exceptions are thrown
inside them, even if they are captured.

closes #331

### CODE REVIEW

**Comment by thepiwo** [comment]
sounds like a good approach to me, I can't judge the code

**Review by thepiwo — APPROVED** [review]
cool!

**Review by jyeshe — COMMENTED** [review]
Neat code for a insightful idea. I would only use behaviours instead of protocols as they are designed to convey the idea of a set of functions that needs to be implemented.

A part of that, I couldn't see any failure. The existing code looks was very mapped to mutations. 

I would only wait for the tests after a full sync and merge it after name consistency is solved as proposed in the meeting.

File: lib/ae_mdw/db/sync/transaction.ex
Comment (jyeshe) [inline lib/ae_mdw/db/sync/transaction.ex:114]: Actually as a matter of practical result, in order to avoid the exceptions inside mnesia transactions, we should probably replace the banging functions behaviour of throwing/raising errors with an `:mnesia.abort`.

Comment (sborrazas) [reply lib/ae_mdw/db/sync/transaction.ex:114]: A few things regarding these comment:
* The `ok!` is outside of a mnesia transaction, so there's no need to do any `:mnesia.abort`
* Any exception that is never supposed to happen shouldn't be handled gracefully, defensive code would just contaminate the codebase
* This PR does not fix exceptions that might happen inside the mnesia transaction that are never caught (those work just fine). This PR is about removing the exceptions that get caught inside the mnesia transaction.

File: lib/ae_mdw/db/mutation.ex
Comment (sborrazas) [reply lib/ae_mdw/db/mutation.ex:16]: A few questions regarding these comments: 

1. How would you implement this with callbacks? E.g. when having a list of mutation structs how do you know which struct should use which callback module?
2. How could `new` be part of a protocol/callback if all `new` functions receive different number and types of arguments?

Perhaps these questions can be best answered by providing some sample pseudo-code

File: lib/ae_mdw/db/model.ex
Comment (sborrazas) [reply lib/ae_mdw/db/model.ex:37]: These two are distinguished by using `Model.tx()` and `Txs.tx()`, we don't need to many every type throughout the codebase differently. All the `Model.x` types are all records, which is easy to distinguish from any other type

Comment (jyeshe) [reply lib/ae_mdw/db/model.ex:37]: There could be types for some indexes (: there are some big ones.... but no big deal. We can expect `Model.x` to be records.

File: lib/ae_mdw/db/mutation.ex
Comment (jyeshe) [reply lib/ae_mdw/db/mutation.ex:16]: To define async tasks which are called by consumers, we have this callback:
```
  @callback process(args :: list()) :: :ok
```
where the workers/processor pattern-matches the specific list like:
```
def process([contract_pk]) do
or
def process([contract_pk, kbi, mbi, create_txi]) do
...
```

Comment (jyeshe) [reply lib/ae_mdw/db/mutation.ex:16]: But for `mutate` action we would need a generic Mutation type so for this reason I liked your choice in favor protocols to the monad-like behaviour making things more explicit for now.

File: lib/ae_mdw/db/sync/transaction.ex
Comment (jyeshe) [reply lib/ae_mdw/db/sync/transaction.ex:114]: Yeah, I have marked the wrong `ok!`, my bad. I meant one inside the mutations that are executed inside the transaction like this:
https://github.com/aeternity/ae_mdw/blob/61b20faabaf9a83de9dd84650b2c0c6c65f31521/lib/ae_mdw/db/sync/name.ex#L116

The proposal would be to replace the use of banging functions like `cache_through_read!` with their non banging counterparts and the nil case should abort a transaction. It would be a complement to the same good idea of the PR, to decrease or eliminate as much as possible the exceptions inside the `:mnesia.transaction`.

**Comment by jyeshe** [comment]
Despite changing the Mdw core responsibility, looks trustworthy. Tests in progress with some bugfixes but related only to rendering so far (unrelated to the PR).

File: lib/ae_mdw/db/sync/transaction.ex
Comment (sborrazas) [reply lib/ae_mdw/db/sync/transaction.ex:114]: These kind of exceptions where the code would blow up if there's a problem are not problematic for mnesia. If the exception isn't captured, mnesia will abort the transaction and no strange errors will appear. 

The problem with exceptions in mnesia are those exceptions that are captured inside the transaction function

**Review by jyeshe — APPROVED** [review]

**Pull request merged**
